# **📖기초 라이브러리 사용 설명서 1탄**

1. **Numpy**

2. **np.array()**

  1) ndarray의 데이터 타입

  2) ndarray를 편리하게 생성하기

3.  **차원과 크기 변경: reshape()**

4.  **데이터 세트 선택하기 - 인덱싱(Indexing)**

5. **행렬의 정렬 - sort()와 argsort()**

6. **선형대수 연산 - 행렬 내적과 전치 행렬 구하기**

7. **배열 간 사칙연산**

8. **Broadcasting**

9. **집계(Aggregation)**

  1) 합계

  2) 평균

  3) 곱셈

  4) 최댓값과 최솟값

  5) 표준편차와 분산


<div><span>



---



## **1. Numpy**

선형대수는 머신러닝 주요 알고리즘 중 하나입니다. 
<br/>넘파이는 파이썬에서 선형대수 기반 프로그램을 가능하게 하는 대표적인 패키지이며 빠른 계산과 배열 연산 능력을 보장합니다.

수행 성능이 중요시될 때는 C/C++ 기반의 코드를 넘파이에서 호출하여 쉽게 통합할 수도 있습니다.
<br/> 딥러닝 프레임워크인 텐서플로가 앞서 말한 방식으로 작성됐습니다.

넘파이는 다양한 데이터 핸들링 기능도 제공하지만 데이터 처리 패키지인 판다스의 편리성은 따라가지 못합니다. 
<br/> 그러나 많은 머신러닝 알고리즘이 넘파이를 기반으로 하기 때문에 넘파이에 익숙해지는 것이 매우 중요합니다.

그럼 지금부터 코드를 하나씩 따라 작성하며 넘파이에 대해 배워보겠습니다!😉

ㅤ

먼저, 넘파이 모듈을 임포트 합니다.

In [ ]:
import numpy as np

ㅤ

## **2. np.array()**

*   리스트와 같은 다양한 인자를 입력 받아서 ndarray로 변환하는 기능을 수행합니다.
*   .shape로 배열의 모양을 확인할 수 있으며 ndarray의 행과 열의 수를 튜플 형태로 가집니다.
*   대괄호[] 중첩으로 배열이 증가합니다.






ㅤ

In [ ]:
#1차원 배열
배열1 = np.array([1, 2, 3])
print(배열1)
print("배열1의 array 형태: ", 배열1.shape)

#2차원 배열
배열2 = np.array([[1, 2, 3], [4, 5, 6]])
print(배열2)
print("배열2의 array 형태: ", 배열2.shape)


#3차원 배열
배열3 = np.array([[[1, 2, 3], [4, 5, 6], [7, 8, 9]]])
print(배열3)
print("배열3의 array 형태: ", 배열3.shape)

[1 2 3]
배열1의 array 형태:  (3,)
[[1 2 3]
 [4 5 6]]
배열2의 array 형태:  (2, 3)
[[[1 2 3]
  [4 5 6]
  [7 8 9]]]
배열3의 array 형태:  (1, 3, 3)


배열 1은 1차원 array로 3개의 데이터를 가지고 있음을 뜻하고 배열2는 2차원 array로 2개의 행과 3개의 열을 가지고 있음을 뜻합니다.

배열3은 3차원 array로 1개의 행과 3개의 열을 가진 배열이 3개가 있음을 뜻합니다.

ㅤ

### **1) ndarray의 데이터 타입**






숫자값, 문자열 값, bool 값 등이 모두 가능합니다.

연산의 특성상 같은 데이터 타입만 가능합니다. 즉, 한 개의 ndarray 객체에 int와 float가 함께 있을 수 없습니다. 

dtype으로 데이터 타입 속성을 확인 할 수 있습니다.

In [ ]:
list1 = [1.5, 2.4, 3.7]
print(type(list1))

array1 = np.array(list1)
print(type(array1))

print(array1, array1.dtype)



<class 'list'>
<class 'numpy.ndarray'>
[1.5 2.4 3.7] float64


리스트 자료형은 list1은 float 숫자인 1.5, 2.4, 3.7을 값으로 가지고 있으며, ndarray로 변경할 수 있습니다. 

변경된 ndarray 내의 데이터값은 모두 float64형 입니다.








ㅤ

In [ ]:
list2 = [1, 2, 'test']
array2 = np.array(list2)
print(array2, array2.dtype)

['1' '2' 'test'] <U21


int형 값과 문자열이 섞여 있는 리스트를 ndarray로 변환한 array2는 숫자형 값이 더 큰 데이터 타입을 가진 문자열 값으로 변환됩니다. 

서로 다른 데이터 타입이 섞여 있을 경우 데이터 타입이 더 큰 데이터 타입으로 변환됩니다. 


> int형: 8bit, 16bit, 32bit


> unsigned int형: 8bit, 16bit, 32bit


> float형: 16bit, 32bit, 64bit, 128bit











ㅤ

메모리를 절약해야 할 때 astype()을 사용하여 ndarray 내 데이터값의 타입을 변경할 수 있습니다.

In [ ]:
array_int = np.array([1, 2, 3])
print(array_int, array_int.dtype)

array_float = array_int.astype('float64')
print(array_float, array_float.dtype)

[1 2 3] int64
[1. 2. 3.] float64


ㅤ

### **2) ndarray를 편리하게 생성하기**


**[ arange() ]**

arange()는 표준 함수 range()와 유사합니다.

0부터 함수 인자 값 -1까지의 값을 순차적으로 ndarray의 데이터값으로 변환합니다.

In [ ]:
sequence_array = np.arange(10)
print(sequence_array)

[0 1 2 3 4 5 6 7 8 9]


ㅤ

**[ zeros() ]**

zeros()는 모든 값을 0으로 채운 배열을 만듭니다.

In [ ]:
zeros1 = np.zeros((3, 2))
print(zeros1)

[[0. 0.]
 [0. 0.]
 [0. 0.]]


In [ ]:
zeros2 = np.zeros([3, 2])
print(zeros2)

[[0. 0.]
 [0. 0.]
 [0. 0.]]


ㅤ

**[ ones() ]**

ones()는 모든 값을 1로 채운 배열을 만듭니다.

In [ ]:
ones1 = np.ones((3, 2))
print(ones1)

[[1. 1.]
 [1. 1.]
 [1. 1.]]


In [ ]:
ones2 = np.ones([3, 2])
print(ones2)

[[1. 1.]
 [1. 1.]
 [1. 1.]]


ㅤ

**[ random.randn() ]**

random.randn()는 랜덤한 숫자 배열을 만듭니다.

In [ ]:
rd = np.random.randn(3, 2) #중첩 괄호는 사용할 수 없습니다.
print(rd)

[[ 0.01479822 -1.3680088 ]
 [-0.8747457   0.15077185]
 [ 1.30192279  0.36353498]]


ㅤ

## **3. 차원과 크기 변경: reshape()**

ndarray를 특정 차원 및 크기로 변환합니다.

지정된 사이즈로 변경이 불가능하면 오류가 발생합니다.

In [ ]:
array1 = np.arange(10)
print("array1:\n", array1)

array2 = array1.reshape(2, 5) 
print("array2:\n", array2)

array1:
 [0 1 2 3 4 5 6 7 8 9]
array2:
 [[0 1 2 3 4]
 [5 6 7 8 9]]


ㅤ

reshape(-1, 1)을 이용해 3차원을 2차원으로 2차원을 1차원으로 변경할 수 있습니다.

In [ ]:
array1 = np.arange(8)

array3d = array1.reshape((2, 2, 2))
print('array3d: \n', array3d)

array2d = array3d.reshape(-1, 1)
print("array2d:\n", array2d)


array3d: 
 [[[0 1]
  [2 3]]

 [[4 5]
  [6 7]]]
array2d:
 [[0]
 [1]
 [2]
 [3]
 [4]
 [5]
 [6]
 [7]]


ㅤ

## **4. 데이터 세트 선택하기: 인덱싱(Indexing)**

### **1) 단일 값 추출**

객체에 해당하는 위치의 인덱스 값을 [ ]안에 입력합니다.

In [ ]:
array1 = np.arange(start=1, stop=10)
print('array1:', array1)

print('array1[2]:', array1[2])

print('맨 뒤의 값: ', array1[-1], '\n맨 뒤에서 두 번째 값: ', array1[-2])

array1: [1 2 3 4 5 6 7 8 9]
array1[2]: 3
맨 뒤의 값:  9 
맨 뒤에서 두 번째 값:  8


ㅤ

단일 인덱스를 이용해 데이터값 수정도 가능합니다.

In [ ]:
array1[0] = 9
print(array1)

[9 2 3 4 5 6 7 8 9]


ㅤ

다차원에서 단일 값을 추출할 수 있습니다.

In [ ]:
array2 = array1.reshape(3, 3)
print(array2)

print('1행 1열 값: ', array2[0, 0])
print('1행 2열 값: ', array2[0, 1])

[[9 2 3]
 [4 5 6]
 [7 8 9]]
1행 1열 값:  9
1행 2열 값:  2


ㅤ

In [ ]:
#4차원 배열

array4 = np.random.randn(2, 2, 4, 5).round(3) #5개씩 1차원 배열이 4개, 1차원 배열 4개가 모인 2차원 배열이 2개, 2차원 배열 2개가 있는 3차원 배열이 2개
print(array4)

print("-"*45)

print('\n array4[0, 0]: \n', array4[0, 0])
print('\n array4[0, 1]: \n', array4[0, 1])
print('\n array4[1, 0]: \n', array4[1, 0])
print('\n array4[1, 0]: \n', array4[1, 1])

print("-"*45)

print('\n array4[0][1][3]: \n', array4[0][1][3])
print('\n array4[0][1][3][4]: \n', array4[0, 1, 3, 4]) #대괄호 대신 쉼표로도 차원 접근이 가능합니다.

[[[[-0.742 -1.657  0.178  0.221 -2.013]
   [ 1.071 -0.725 -2.689 -2.062  1.089]
   [-0.54  -0.157 -1.44   2.468 -0.307]
   [-0.059  0.687 -0.034  1.387 -1.435]]

  [[ 1.356  1.52   0.402  1.286  1.316]
   [ 0.244 -2.103  2.132 -0.633  0.224]
   [-0.5   -1.684 -1.674 -1.461  0.265]
   [ 0.269 -1.83  -0.77  -0.179 -0.29 ]]]


 [[[-0.886  1.901  1.032  1.703  0.633]
   [ 1.656 -1.171 -0.617 -0.751 -0.052]
   [-0.792  0.442  0.038 -0.089  1.099]
   [-0.332  0.079  2.297 -0.18   0.175]]

  [[-1.311 -0.962 -0.328  1.132  1.323]
   [ 1.621 -1.786 -0.205 -0.993  0.19 ]
   [-0.647  1.129  0.243  2.171 -0.172]
   [ 2.135 -1.206 -0.575 -1.409 -0.227]]]]
---------------------------------------------

 array4[0, 0]: 
 [[-0.742 -1.657  0.178  0.221 -2.013]
 [ 1.071 -0.725 -2.689 -2.062  1.089]
 [-0.54  -0.157 -1.44   2.468 -0.307]
 [-0.059  0.687 -0.034  1.387 -1.435]]

 array4[0, 1]: 
 [[ 1.356  1.52   0.402  1.286  1.316]
 [ 0.244 -2.103  2.132 -0.633  0.224]
 [-0.5   -1.684 -1.674 -1.461  0.265]


ㅤ

### **2) 슬라이싱(Slicing)**

: 를 이용하여 연속 데이터를 추출합니다.

In [ ]:
array1 = np.arange(start=1, stop = 10)
print(array1[0:2])

[1 2]




1.   : 기호 앞에 시작 인덱스를 생략하면 자동으로 맨 처음 인덱스인 0으로 간주합니다.
2.   : 기호 뒤에 종료 인덱스를 생략하면 자동으로 맨 마지막 인덱스로 간주합니다.
3.   : 기호 앞/뒤에 시작/종료 인덱스를 생략하면 자동으로 맨 처음/맨 마지막 인덱스로 간주합니다.



ㅤ

In [ ]:
print(array1,'\n')
print(array1[:5])
print(array1[3:])
print(array1[:])

[1 2 3 4 5 6 7 8 9] 

[1 2 3 4 5]
[4 5 6 7 8 9]
[1 2 3 4 5 6 7 8 9]


2차원 ndarray에서 슬라이싱으로 데이터에 접근할 때는 콤마(,)로 행과 열 인덱스를 지칭합니다.

ㅤ

In [ ]:
array1 = array1.reshape(3,3)
print(array1)
print('-'*10)

print(array1[0:2])


[[1 2 3]
 [4 5 6]
 [7 8 9]]
----------
[[1 2 3]
 [4 5 6]]


2차원 배열안에 있는 1차원 배열들 중에서 0번째와 1번째 배열을 가져옵니다.

ㅤ

In [ ]:
print(array1)
print('-'*10)

print(array1[0:3, 1:3])


[[1 2 3]
 [4 5 6]
 [7 8 9]]
----------
[[2 3]
 [5 6]
 [8 9]]


2차원 배열안에 있는 1차원 배열들 중에서 0번째부터 2번째까지 가져오고 가져온 1차원 배열 안에서 1번째부터 2번째까지 가져옵니다.

ㅤ

In [ ]:
print(array1)
print('-'*10)

print(array1[1:, :2])

[[1 2 3]
 [4 5 6]
 [7 8 9]]
----------
[[4 5]
 [7 8]]


2차원 배열안에 있는 1차원 배열들 중에서 1번째부터 맨 마지막번째까지 가져오고 가져온 1차원 배열안에서 맨 처음부터 1번째까지 가져옵니다.

ㅤ

In [ ]:
print(array1)
print('-'*10)

print(array1[:3, 0])

[[1 2 3]
 [4 5 6]
 [7 8 9]]
----------
[1 4 7]


2차원 배열안에 있는 1차원 배열들 중에서 0번째부터 2번째까지 가져오고 가져온 1차원 배열안에서 0번째까지 가져옵니다. 

ㅤ

### **3) 팬시 인덱싱**





리스트나 ndarray로 인덱스 집합을 지정하면 해당 위치의 인덱스에 해당하는 ndarray를 반환합니다.

In [ ]:
print(array1)
print('-'*10)

print(array1[[0,2]])
print('-'*10)

print(array1[[0,2], 1])
print('-'*10)

print(array1[[0,2], 0:2])




[[1 2 3]
 [4 5 6]
 [7 8 9]]
----------
[[1 2 3]
 [7 8 9]]
----------
[2 8]
----------
[[1 2]
 [7 8]]


ㅤ

### **4) 불린 인덱싱**

조건 필터링과 검색을 동시에 할 수 있기 때문에 매우 자주 사용됩니다.

In [ ]:
boolean = array1[array1>3]

print('array1 > 3인 불린 인덱싱 결과 값:',  boolean)

print('객체반환: \n', array1>3)

array1 > 3인 불린 인덱싱 결과 값: [4 5 6 7 8 9]
객체반환: 
 [[False False False]
 [ True  True  True]
 [ True  True  True]]


ㅤ

## **5. 행렬의 정렬  - sort()와 argsort()**

### **1) 행렬 정렬**



*   np.sort() : 원본 행렬은 그대로 유지한 채 원 행렬의 정렬된 행렬을 반환합니다.
*   ndarray.sort() : 원본 행렬 자체를 정렬합니다. 반환 값은 None 입니다. 



In [ ]:
org_array = np.array([4, 2, 8, 6])

print('원본 행렬: ', org_array)
print('np.sort 호출 후 반환된 행렬: ', np.sort(org_array))
print('np.sort 호출 후 원본 행렬: ', org_array)
print('ndarray.sort 호출 후 반환된 행렬:', org_array.sort())
print('ndarray.sort 호출 후 원본 행렬:', org_array)

원본 행렬:  [4 2 8 6]
np.sort 호출 후 반환된 행렬:  [2 4 6 8]
np.sort 호출 후 원본 행렬:  [4 2 8 6]
ndarray.sort 호출 후 반환된 행렬: None
ndarray.sort 호출 후 원본 행렬: [2 4 6 8]


ㅤ

내림차순으로 반환하기 위해서는 np.sort에서 다음과 같이 [::-1]을 붙여줍니다.

In [ ]:
print('내림차순 정렬: ', np.sort(org_array)[::-1])

내림차순 정렬:  [8 6 4 2]


ㅤ

2차원 배열은 axis 축 값 설정을 통해 행열 정렬을 수행합니다. 

*   axis = 0 : 행 방향
*   axis = 1 : 열 방향



In [ ]:
array2 = np.array([[8, 1], [7, 12]])

print(array2)

print('열 방향 정렬:\n', np.sort(array2, axis = 1))
print('행 방향 정렬:\n', np.sort(array2, axis = 0))

[[ 8  1]
 [ 7 12]]
열 방향 정렬:
 [[ 1  8]
 [ 7 12]]
행 방향 정렬:
 [[ 7  1]
 [ 8 12]]


ㅤ

### **2) np.argsort() : 정렬된 행렬의 인덱스를 반환하기**

argsort()는 넘파이에서 매우 활용도가 높습니다. 

In [ ]:
org_array = np.array([4, 2, 8, 6])

print('np.argsort 호출 후 반환된 인덱스:', np.argsort(org_array))


np.argsort 호출 후 반환된 인덱스: [1 0 3 2]


ㅤ

np.argsort 역시 np.sort와 마찬가지로 내림차순 반환이 가능합니다.

In [ ]:
print('np.argsort 호출 후 반환된 인덱스:', np.argsort(org_array)[::-1])

np.argsort 호출 후 반환된 인덱스: [2 3 0 1]


ㅤ

## **6. 선형대수 연산 - 행렬 내적과 전치 행렬 구하기**

### **1) 행렬 내적(행렬 곱)**

np.dot()을 이용해 계산이 가능합니다.

In [ ]:
a = np.array([[1, 2, 3], [4, 5, 6]])
print('a행렬: \n', a)

b = np.array([[7, 8], [9, 10], [11, 12]])
print('b행렬: \n', b)

print('a와 b의 행렬내적 결과:\n', np.dot(a,b))

a행렬: 
 [[1 2 3]
 [4 5 6]]
b행렬: 
 [[ 7  8]
 [ 9 10]
 [11 12]]
a와 b의 행렬내적 결과:
 [[ 58  64]
 [139 154]]




*   [1, 2, 3]*[7, 9, 11] = 1 * 7 + 2 * 9 + 3* 11 = 58
*   [1, 2, 3]*[8, 10, 12] = 1 * 8 + 2 * 10 + 3 * 12 = 64





*   [4, 5, 6]*[7, 9, 11] = 4 * 7 + 5 * 9 + 6 * 11 = 139
*   [4, 5, 6]*[8, 10, 12] = 4 * 8 + 5 * 10 + 6 * 12 = 154



ㅤ

### **2) 전치 행렬**

*  원 행렬에서 행과 열의 위치를 교환한 원소로 구성한 행렬
*  np.transpose()를 사용합니다.

In [ ]:
a = np.array([[1, 2], [3, 4]])
print('a행렬:\n', a)
print('a의 전치행렬:\n', np.transpose(a))

a행렬:
 [[1 2]
 [3 4]]
a의 전치행렬:
 [[1 3]
 [2 4]]


ㅤ

## **7. 배열 간 사칙연산**

동일한 모양의 두 배열을 사칙연산합니다.

In [ ]:
a = np.array([[11, 12, 13], [2, 3, 4]])
print('a행렬:\n', a)
print('a+a =\n', a+a)

a행렬:
 [[11 12 13]
 [ 2  3  4]]
a+a =
 [[22 24 26]
 [ 4  6  8]]


배열 내 같은 위치의 원소끼리 연산을 합니다.

ㅤ

## **8. Broadcasting**

배열과 1개의 숫자를 사칙 연산합니다.

In [ ]:
print('a행렬:\n', a)
print('a+13=\n', a+13)
print('a-13=\n', a-13)
print('a*3=\n', a*3)


a행렬:
 [[11 12 13]
 [ 2  3  4]]
a+13=
 [[24 25 26]
 [15 16 17]]
a-13=
 [[ -2  -1   0]
 [-11 -10  -9]]
a*3=
 [[33 36 39]
 [ 6  9 12]]


ㅤ

같은 수의 행과 열을 가진 여러 개의 배열을 연산할 수도 있습니다.

In [ ]:
a1 = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
a2 = np.array([[1, 2, 3], [1, 2, 3], [1, 2, 3]])

print(a1, '\n+\n', a2, '\n=', a1+a2)

[[1 2 3]
 [4 5 6]
 [7 8 9]] 
+
 [[1 2 3]
 [1 2 3]
 [1 2 3]] 
= [[ 2  4  6]
 [ 5  7  9]
 [ 8 10 12]]


ㅤ

다른 형태의 배열끼리도 연산이 가능합니다.

In [ ]:
broad = np.arange(10).reshape(5, 2)
casting = np.arange(10, 12)

print(broad)
print('-'*10)
print(casting)
print('-'*10)
print(broad+casting)

[[0 1]
 [2 3]
 [4 5]
 [6 7]
 [8 9]]
----------
[10 11]
----------
[[10 12]
 [12 14]
 [14 16]
 [16 18]
 [18 20]]


ㅤ

## **9. 집계(Aggregation)**

### **1) 합계**

배열의 합계는 .sum() 을 이용합니다.

In [ ]:
print('a행렬:\n', a)
print(a.sum())

a행렬:
 [[11 12 13]
 [ 2  3  4]]
45


ㅤ

축을 설정하여 열 방향 혹은 행 방향으로만 더할 수 있습니다.

In [ ]:
print('열 방향 합계: ', a.sum(axis=1))
print('행 방향 합계: ', a.sum(axis=0))




열 방향 합계:  [36  9]
행 방향 합계:  [13 15 17]


ㅤ

### **2) 평균**

배열의 평균은 .mean()을 이용합니다.

In [ ]:
print(a.mean())

7.5


ㅤ

### **3) 곱셈**

배열의 곱샘은 .prod()을 이용합니다.

In [ ]:
print(a.prod())

41184


ㅤ

### **4) 최댓값과 최솟값**

최댓값과 최솟값은 각각 .max()와 .min()을 이용합니다.

In [ ]:
print(a.max())
print(a.min())


13
2


ㅤ

인덱싱으로 해당 열의 최댓값과 최솟값을 구할 수도 있습니다. 

In [ ]:
print(a)
print(a[1].max())
print(a[0].min())

[[11 12 13]
 [ 2  3  4]]
4
11


ㅤ

.argmax()와 .argmin()을 이용하여 배열에서 최댓값인 원소의 인덱스 번호와 최솟값인 원소의 인덱스 번호를 구할 수 있습니다.

In [ ]:
print(a.argmax())
print(a.argmin())

2
3


ㅤ

### **5) 표준편차와 분산**

표준편차와 분산은 각각 .std()와 .var()을 이용합니다. 

In [ ]:
print(a.std())
print(a.var())

4.573474244670748
20.916666666666668



---




이렇게 Numpy에 대해 공부해봤습니다!

배열간 연산과 인덱싱, 슬라이싱은 Pandas와 함께 데이터 병합과 전처리를 진행할 때 유용하게 쓰이는 방법입니다. 

자주 사용하는 만큼 중요한 라이브러리이기 때문에 같이 코드를 따라 작성해보며 공부하시면 데이터 분석을 하는데 도움이 되실 거예요!

<br>📣 다음은 **Pandas를 다루는 기초 라이브러리 사용 설명서 2탄**이 업로드될 예정입니다!


많은 관심과 애정 부탁드려요💛 감사합니다!
